### PR TITLE
Run3

### PR DESCRIPTION
**Pull request by MatthewDittrich**
- Adding 2022 Btagging JSON
- Adding 2022 Golden JSON
- Adding some run3 cross section values
- Adding 2022 Lumi values
- Adding era option for creating sample JSONs (due to some era dependency in 2022)

### CODE REVIEW

File: topcoffea/data/goldenJsons/Cert_Collisions2022_355100_362760_Golden.txt
Comment (kmohrman) [inline topcoffea/data/goldenJsons/Cert_Collisions2022_355100_362760_Golden.txt:1]: Could you add to the readme to put a reference of where the 2022 golden json comes from?

File: topcoffea/data/btag_sf_correctionlib/README.md [hunk 1]
```diff
@@ -8,10 +8,14 @@ cp /cvmfs/cms.cern.ch/rsync/cms-nanoAOD/jsonpog-integration/POG/BTV/2016preVFP_U
 cp /cvmfs/cms.cern.ch/rsync/cms-nanoAOD/jsonpog-integration/POG/BTV/2016postVFP_UL/btagging.json.gz 2016postVFP_UL_btagging.json.gz
 cp /cvmfs/cms.cern.ch/rsync/cms-nanoAOD/jsonpog-integration/POG/BTV/2017_UL/btagging.json.gz 2017_UL_btagging.json.gz
 cp /cvmfs/cms.cern.ch/rsync/cms-nanoAOD/jsonpog-integration/POG/BTV/2018_UL/btagging.json.gz 2018_UL_btagging.json.gz
+cp /cvmfs/cms.cern.ch/rsync/cms-nanoAOD/jsonpog-integration/POG/BTV/2022_Summer22/btagging.json.gz 2022_btagging.json.gz
```
Comment (kmohrman) [inline topcoffea/data/btag_sf_correctionlib/README.md:11]: Super minor thing, but just to help organize the readme visually, could you add a space above this line (so there is a space between the R2 and R2 ones)?

File: topcoffea/data/btag_sf_correctionlib/README.md [hunk 2]
```diff
@@ -8,10 +8,14 @@ cp /cvmfs/cms.cern.ch/rsync/cms-nanoAOD/jsonpog-integration/POG/BTV/2016preVFP_U
...
 cp /cvmfs/cms.cern.ch/rsync/cms-nanoAOD/jsonpog-integration/POG/BTV/2017_UL/btagging.json.gz 2017_UL_btagging.json.gz
 cp /cvmfs/cms.cern.ch/rsync/cms-nanoAOD/jsonpog-integration/POG/BTV/2018_UL/btagging.json.gz 2018_UL_btagging.json.gz
+cp /cvmfs/cms.cern.ch/rsync/cms-nanoAOD/jsonpog-integration/POG/BTV/2022_Summer22/btagging.json.gz 2022_btagging.json.gz 
+cp /cvmfs/cms.cern.ch/rsync/cms-nanoAOD/jsonpog-integration/POG/BTV/2022_Summer22EE/btagging.json.gz 2022EE_btagging.json.gz 
 
 gzip -d 2016preVFP_UL_btagging.json.gz
 gzip -d 2016postVFP_UL_btagging.json.gz
 gzip -d 2017_UL_btagging.json.gz
 gzip -d 2018_UL_btagging.json.gz
+gzip -d 2022_btagging.json.gz
```
Comment (kmohrman) [inline topcoffea/data/btag_sf_correctionlib/README.md:18]: Also here same comment as above (line break would help visually, I think).

File: topcoffea/modules/createJSON.py [hunk 3]
```diff
@@ -151,7 +156,8 @@ def main():
     if args.includeLheWgts:
         sampdic['nSumOfLheWeights'] = n_sum_of_lhe_weights
 
-    outname = sample
+    if sample != '':
+        outname = sample
```
Comment (kmohrman) [inline topcoffea/modules/createJSON.py:160]: Else what happens? I.e. what will outcome be in the case when sample is ''?

File: topcoffea/modules/event_selection.py [hunk 4]
```diff
@@ -26,11 +26,13 @@ def passes_trg_inlst(events,trg_name_lst):
...
 #   - Elements are false if they do not pass any of the triggers defined in dataset_dict
 #   - In the case of data, events are also false if they overlap with another dataset
-def trg_pass_no_overlap(events,is_data,dataset,year,dataset_dict,exclude_dict):
+def trg_pass_no_overlap(events,is_data,dataset,year,dataset_dict,exclude_dict,era=None):
 
-    # The trigger for 2016 and 2016APV are the same
+    # The triggers for 2016 and 2016APV are the same
     if year == "2016APV":
         year = "2016"
+    if year == "2022EE":
```
Comment (kmohrman) [inline topcoffea/modules/event_selection.py:34]: But the trigger lists are not the same for 2022 and 2022 EE, right? So I am not sure if I understand the impact this will have. Could you explain a bit about what's happening here?

File: topcoffea/modules/event_selection.py [hunk 5]
```diff
@@ -46,7 +48,10 @@ def trg_pass_no_overlap(events,is_data,dataset,year,dataset_dict,exclude_dict):
     trg_passes = passes_trg_inlst(events,full_trg_lst)
 
     # In case of data, check if events overlap with other datasets
-    if is_data:
+    if is_data and (era is not None):   #Used for potential era dependency in Run3
```
Comment (kmohrman) [inline topcoffea/modules/event_selection.py:51]: I think this would be easier to read if the logic looked like this:
```
if is_data:
    if era is not none:
        do stuff
    else:
         do other stuff
```

File: topcoffea/modules/event_selection.py [hunk 6]
```diff
@@ -46,7 +48,10 @@ def trg_pass_no_overlap(events,is_data,dataset,year,dataset_dict,exclude_dict):
...
+        trg_passes = passes_trg_inlst(events,dataset_dict[year][dataset])
```
Comment (kmohrman) [inline topcoffea/modules/event_selection.py:52]: If you are checking by era in this case, should the passes_trg_inlist also check era here? (Like you do for overlap check?)

File: topcoffea/modules/sample_lst_jsons_tools.py [hunk 7]
```diff
@@ -47,7 +47,7 @@ def replace_xsec_for_dict_of_samples(samples_dict,out_dir):
         replace_val_in_json(path_to_json,"xsec",new_xsec)
 
 # Wrapper for createJSON.py
-def make_json(sample_dir,sample_name,prefix,sample_yr,xsec_name,hist_axis_name,on_das=False,include_lhe_wgts_arr=False):
+def make_json(sample_dir,sample_name,prefix,sample_yr,xsec_name,hist_axis_name,era,on_das=False,include_lhe_wgts_arr=False):
```
Comment (kmohrman) [inline topcoffea/modules/sample_lst_jsons_tools.py:50]: Can you make this optional? I.e. add a default value (None I guess would be consistent with how you're handling it elsewhere).

File: topcoffea/modules/utils.py [hunk 8]
```diff
@@ -123,6 +123,7 @@ def get_info(fname, tree_name = "Events"):
     raw_events = 0  # The raw number of entries as reported by TTree.num_entries
     gen_events = 0  # Number of gen events according to 'genEventCount' or set to raw_events if not found
     sow_events = 0  # Sum of weights
+    sow_lhe_wgts = 0 #Sume of LHE weights
```
Comment (kmohrman) [inline topcoffea/modules/utils.py:126]: Could you explain a bit about what this is doing and/or why it's needed?

File: topcoffea/params/xsec.yml [hunk 9]
```diff
@@ -91,7 +91,7 @@ ZZTo2Q2Nu               : 4.04
 ZZTo4L                  : 1.256
 
 ggToZHToHToWWTo2L2Nu    : 0.00275
-qqToZHToZTo2L           : 0.0018639
+ggToZHToZTo2L           : 0.0018639
```
Comment (kmohrman) [inline topcoffea/params/xsec.yml:94]: Was this change intentional? Was it a typo in the previous version? Do you have a ref for the xsec?

File: topcoffea/params/xsec.yml [hunk 10]
```diff
@@ -120,11 +120,34 @@ TWZToLL_thad_Wlept  : 0.003004
 TWZToLL_tlept_Whad  : 0.003004
 TWZToLL_tlept_Wlept : 0.0015
 
-# Number from CMS AN-19-144
-tWll : 0.01123
```
Comment (kmohrman) [inline topcoffea/params/xsec.yml:124]: Why getting rid of this?

File: topcoffea/params/xsec.yml [hunk 11]
```diff
@@ -120,11 +120,34 @@ TWZToLL_thad_Wlept  : 0.003004
...
+GluGlutoContinto2Zto4Mu_13p6TeV        : 0.00303575
+GluGlutoContinto2Zto4E_13p6TeV         : 0.00305851
+GluGlutoContinto2Zto4Tau_13p6TeV       : 0.00303575
+
+#Numbers from SMP-24-005
+VH_HtoNonbb_13p6TeV                    : 0.9561
+WZZ_13p6TeV                            : 0.06
+ZZZ_13p6TeV                            : 0.02
+
+data                           : 1.0
```
Comment (kmohrman) [inline topcoffea/params/xsec.yml:153]: Should not need data in more than one place

File: topcoffea/modules/createJSON.py (hunk 3 above)
Comment (MatthewDittrich) [reply topcoffea/modules/createJSON.py:160]: Then the outname would be the specified outname from the argparse. If both have been left empty, then the output would be '.json'
Could we raise an exception if both are empty?

File: topcoffea/modules/event_selection.py (hunk 4 above)
Comment (MatthewDittrich) [reply topcoffea/modules/event_selection.py:34]: I think we should leave this: the trigger list is the same, the problem with 2022 & 2022EE is the sudden change from SIngleMuon and DoubleMuon to just Muon. However, Muon still consists of the triggers from SingleMuon and DoubleMuon. And the triggers for EGamma and MuonEG are all the same

File: topcoffea/modules/event_selection.py (hunk 6 above)
Comment (MatthewDittrich) [reply topcoffea/modules/event_selection.py:52]: No, because the total trigger list will be the same for each era

File: topcoffea/modules/utils.py (hunk 8 above)
Comment (MatthewDittrich) [reply topcoffea/modules/utils.py:126]: This is needed for data files. Without this, the method would try to return sow_lhe_wgts without any declaration. Other option is to have conditional return statements based on is_Data

File: topcoffea/params/xsec.yml (hunk 10 above)
Comment (MatthewDittrich) [reply topcoffea/params/xsec.yml:124]: Mistake

File: topcoffea/params/xsec.yml (hunk 9 above)
Comment (MatthewDittrich) [reply topcoffea/params/xsec.yml:94]: definitely not intentional. Not sure when this happened. Maybe a typo in the previous version that has since been updated and I pushed the old?

File: topcoffea/params/xsec.yml (hunk 11 above)
Comment (MatthewDittrich) [reply topcoffea/params/xsec.yml:153]: I think I removed it from above and added it at the bottom (just to keep it at the bottom for consistency). there should be only one

File: topcoffea/data/goldenJsons/Cert_Collisions2022_355100_362760_Golden.txt
Comment (kmohrman) [reply topcoffea/data/goldenJsons/Cert_Collisions2022_355100_362760_Golden.txt:1]: Thanks for adding the link to where the golden json can be copied from. Sorry to request other changes, but I think these would be helpful:
* A link to the twiki where the lumi information can be found
* Looking at the draft readme, I think it would really improve the readability of this readme to keep the R2 and R3 parts separate in the readme. Could you make each into their own section?

File: topcoffea/modules/createJSON.py (hunk 3 above)
Comment (kmohrman) [reply topcoffea/modules/createJSON.py:162]: Yes I think it would be good to raise an exception instead of writing to ".json"

File: topcoffea/params/xsec.yml [hunk 12]
```diff
@@ -1,4 +1,4 @@
-# Notes about the numbers:
+# Notes about :/the numbers:
```
Comment (kmohrman) [inline topcoffea/params/xsec.yml:1]: What's up with this ":/"? Maybe vim searching while in insert mode :)

File: topcoffea/modules/utils.py (hunk 8 above)
Comment (kmohrman) [reply topcoffea/modules/utils.py:126]: Ah, ok. Can you fix the formatting of the comment (space between "#" and comment) and "Sume" -> "Sum" typo.

File: topcoffea/modules/event_selection.py (hunk 4 above)
Comment (kmohrman) [reply topcoffea/modules/event_selection.py:35]: Ok, thanks for explaining. For the record, now that I look at this function again, I think it would be better if we handed this year stuff on the users' end where we _call_ this function (e.g. just pass "2022" for the year in the case where the year is "2022EE" instead of having this hard coded in the function itself. 
But I'd say not necessary to fix in this PR. I've made an [issue](https://github.com/TopEFT/topcoffea/issues/24) so that we remember it in the future (if we'd like to address it someday).

**Comment by kmohrman** [comment]
Thank @MatthewDittrich for addressing all of my comments. Please edit the description of the PR (one bullet point for each of the main things you added), then after that, please remove the `[WIP]` from the PR title. Once that's done, please request a review (using the "Reviewers" button on the top right of the PR, screenshot attached) from @bryates and @Andrew42. 
Screenshot 2024-04-02 at 3 26 22 PM

**Review by bryates — CHANGES_REQUESTED** [review]

File: topcoffea/modules/createJSON.py [hunk 13]
```diff
@@ -60,6 +62,8 @@ def main():
     sampdic['treeName']     = treeName
     sampdic['histAxisName'] = histAxisName
     sampdic['options']      = options
+    if era is not None:
```
[the remark this answers is not in the document]
Comment (kmohrman) [reply topcoffea/modules/createJSON.py:65]: I agree that sounds like a good check to include.

File: topcoffea/modules/event_selection.py [hunk 14]
```diff
@@ -47,8 +50,12 @@ def trg_pass_no_overlap(events,is_data,dataset,year,dataset_dict,exclude_dict):
 
     # In case of data, check if events overlap with other datasets
     if is_data:
-        trg_passes = passes_trg_inlst(events,dataset_dict[year][dataset])
-        trg_overlaps = passes_trg_inlst(events, exclude_dict[year][dataset])
+        if era is not None: # Used for potential era dependency in Run3
+            trg_passes = passes_trg_inlst(events,dataset_dict[year][dataset])
```
[the remark this answers is not in the document]
Comment (kmohrman) [reply topcoffea/modules/event_selection.py:54]: There is already era dependence for R3. @MatthewDittrich I think it might be more clear to just remove the word "potential" from the comment

[the remark this answers is not in the document]
Comment (bryates) [reply topcoffea/modules/event_selection.py:54]: Oh yes, I messed `[era]` in the next line.

File: topcoffea/modules/createJSON.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (bryates) [reply topcoffea/modules/createJSON.py:65]: @kmohrman do you have a preference on a warning vs an exception? I see @MatthewDittrich added a print statement for a warning. If I remember correctly, this script prints out a bunch of stuff, so I'm worried the warning might get buried.

[the remark this answers is not in the document]
Comment (kmohrman) [reply topcoffea/modules/createJSON.py:65]: @bryates I think for me, it seems a warning is ok. It is conceivable that someone might want to make a R3 json that does not need the era listed; the era is only needed for overlap removal. So if e.g. you were just making one json for one dataset (for some test) you would not need the era included since you would not be doing any overlap removal.

[the remark this answers is not in the document]
Comment (bryates) [reply topcoffea/modules/createJSON.py:65]: Ok good point. I wasn't sure if the era was necessary or just an option. This looks good then.

File: topcoffea/modules/event_selection.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (MatthewDittrich) [reply topcoffea/modules/event_selection.py:54]: Yes there is era dependency for sure. I removed the word 'potential' to indicate this

**Review by bryates — APPROVED** [review]

**Comment by bryates** [comment]
> Ok, looks like all of the comments have been addressed. @bryates are you happy for it to be merged?

Yup, I'm happy with this!

**Comment by kmohrman** [comment]
Ok, thanks for the review! I'll go ahead and merge it now.
